libpdbg/device.c

```diff
@@ -48,11 +48,118 @@ static const char *take_name(const char *name)
 	return name;
 }
 
+static const char *get_unitname(const struct pdbg_target *node)
+{
+	const char *c = strchr(node->dn_name, '@');
+
+	if (!c)
+		return NULL;
+
+	return c + 1;
+}
+
+/*If a string contains alphanumeric characters then it extracts 
+	the number from that string (from the end only) and 
+	returns the extracted number and the chopped off string*/
+static unsigned int strip_num_at_the_end(char *str) 
+{
+	unsigned int only_number;
+	int len = strlen(str);
+	char new_num_str[len];
+
+	int i;
+	/*Traverse from the end of the string looking for alphabet after the digit */
+	for(i = (len - 1); i >= 0; i--) {
+		if(!isdigit(str[i])) {
+		/* This is the first character from the end. So 
+		copy the number starting from the next index */
+		strcpy(new_num_str, &str[i+1]);
+		break;
+		}
+	}
+	str[i+1] = '\0';
+	only_number = atoi(new_num_str);
+	return only_number;
+}
+
+/*Compare the pdbg_targets according to their name. Initially checks where it falls 
+as per alphabetical order. Then from there use the number at the end of the name to 
+sort numerically within that group
+e.g. perv3 shall be placed after omi0 and before perv10 */
+static int dt_cmp_subnodes(const struct pdbg_target *a, const struct pdbg_target *b)
+{
+	const char *a_unit = get_unitname(a);
+	const char *b_unit = get_unitname(b);
+	ptrdiff_t basenamelen = a_unit - a->dn_name;
+	/* sort hex unit addresses by number */
+	if (a_unit && b_unit && !strncmp(a->dn_name, b->dn_name, basenamelen)) {
+		unsigned long long a_num, b_num;
+		char *a_end, *b_end;
+
+		a_num = strtoul(a_unit, &a_end, 16);
+		b_num = strtoul(b_unit, &b_end, 16);
+
+		/* only compare if the unit addr parsed correctly */
+		if (*a_end == 0 && *b_end == 0)
+			return (a_num > b_num) - (a_num < b_num);
+	}
+	char a_name_without_num[strlen(a->dn_name) + 1];
+	char b_name_without_num[strlen(b->dn_name) + 1];
+	unsigned int a_only_num = 0, b_only_num = 0;
+
+	/* Create a copy as we are going to chop the number from it and we dont want to update the original string */
+	strcpy(a_name_without_num, a->dn_name);
+	strcpy(b_name_without_num, b->dn_name);
+
+	/* Collect the numbers into variables to compare */
+	a_only_num = strip_num_at_the_end(a_name_without_num);
+	b_only_num = strip_num_at_the_end(b_name_without_num);
+
+	//If both of the targets are of the same group, sort using the number at the end
+	if(strcmp(a_name_without_num, b_name_without_num) == 0 && (a_only_num) && (b_only_num) )
```

Comment:
a_only_num can;t this be zero? for say core0

Reply:
Thank you for bringing up this concern. I will test it out and may be go for a default value of -1 (negative one) if required

Reply:
Those checks are not required. Thank you for catching it.  Though it was working correctly (for now) by adding to the end, removed the checks to ensure it works for all other cases.